src/matrix_cookbook/mat_eq27_lib.lean

```diff
@@ -0,0 +1,214 @@
+begin
+  rw matrix.trace,
+  simp only [diag],
+  rw pow_two, rw mul_eq_mul, rw fin.sum_univ_four,
+  repeat{rw mul_apply, rw fin.sum_univ_four},
+  ring,
+end
+
+lemma trace_pow_three_fin_four {A : matrix (fin 4) (fin 4) R}:
+  trace (A^3) = 
+    A 0 0*(A 0 0^2 + A 0 1*A 1 0 + A 0 2*A 2 0 + A 0 3*A 3 0) + 
+    A 1 1*(A 1 1^2 + A 0 1*A 1 0 + A 1 2*A 2 1 + A 1 3*A 3 1) + 
+    A 2 2*(A 2 2^2 + A 0 2*A 2 0 + A 1 2*A 2 1 + A 2 3*A 3 2) + 
+    A 3 3*(A 3 3^2 + A 0 3*A 3 0 + A 1 3*A 3 1 + A 2 3*A 3 2) + 
+    A 1 0*(A 0 0*A 0 1 + A 0 1*A 1 1 + A 0 2*A 2 1 + A 0 3*A 3 1) + 
+    A 2 0*(A 0 0*A 0 2 + A 0 1*A 1 2 + A 0 2*A 2 2 + A 0 3*A 3 2) + 
+    A 0 1*(A 0 0*A 1 0 + A 1 0*A 1 1 + A 1 2*A 2 0 + A 1 3*A 3 0) + 
+    A 3 0*(A 0 0*A 0 3 + A 0 1*A 1 3 + A 0 2*A 2 3 + A 0 3*A 3 3) + 
+    A 2 1*(A 0 2*A 1 0 + A 1 1*A 1 2 + A 1 2*A 2 2 + A 1 3*A 3 2) + 
+    A 0 2*(A 0 0*A 2 0 + A 1 0*A 2 1 + A 2 0*A 2 2 + A 2 3*A 3 0) + 
+    A 3 1*(A 0 3*A 1 0 + A 1 1*A 1 3 + A 1 2*A 2 3 + A 1 3*A 3 3) + 
+    A 1 2*(A 0 1*A 2 0 + A 1 1*A 2 1 + A 2 1*A 2 2 + A 2 3*A 3 1) + 
+    A 0 3*(A 0 0*A 3 0 + A 1 0*A 3 1 + A 2 0*A 3 2 + A 3 0*A 3 3) + 
+    A 3 2*(A 0 3*A 2 0 + A 1 3*A 2 1 + A 2 2*A 2 3 + A 2 3*A 3 3) + 
+    A 1 3*(A 0 1*A 3 0 + A 1 1*A 3 1 + A 2 1*A 3 2 + A 3 1*A 3 3) + 
+    A 2 3*(A 0 2*A 3 0 + A 1 2*A 3 1 + A 2 2*A 3 2 + A 3 2*A 3 3)
+ := 
+begin
+  rw [pow_three, matrix.trace, fin.sum_univ_four,  
+    mul_eq_mul,  mul_eq_mul], 
+  simp only[diag],
+  repeat{rw mul_apply, rw fin.sum_univ_four},
+  ring,
+end
+
+lemma det_fin_four (A : matrix (fin 4) (fin 4) R) :
+A.det = 
+  A 0 0*A 1 1*A 2 2*A 3 3 - A 0 0*A 1 1*A 2 3*A 3 2 - A 0 0*A 1 2*A 2 1*A 3 3 + 
+  A 0 0*A 1 2*A 2 3*A 3 1 + A 0 0*A 1 3*A 2 1*A 3 2 - A 0 0*A 1 3*A 2 2*A 3 1 -
+  A 0 1*A 1 0*A 2 2*A 3 3 + A 0 1*A 1 0*A 2 3*A 3 2 + A 0 1*A 1 2*A 2 0*A 3 3 - 
+  A 0 1*A 1 2*A 2 3*A 3 0 - A 0 1*A 1 3*A 2 0*A 3 2 + A 0 1*A 1 3*A 2 2*A 3 0 +
+  A 0 2*A 1 0*A 2 1*A 3 3 - A 0 2*A 1 0*A 2 3*A 3 1 - A 0 2*A 1 1*A 2 0*A 3 3 + 
+  A 0 2*A 1 1*A 2 3*A 3 0 + A 0 2*A 1 3*A 2 0*A 3 1 - A 0 2*A 1 3*A 2 1*A 3 0 -
+  A 0 3*A 1 0*A 2 1*A 3 2 + A 0 3*A 1 0*A 2 2*A 3 1 + A 0 3*A 1 1*A 2 0*A 3 2 - 
+  A 0 3*A 1 1*A 2 2*A 3 0 - A 0 3*A 1 2*A 2 0*A 3 1 + A 0 3*A 1 2*A 2 1*A 3 0 := 
+begin
+  -- There is no way but to brute force this. "Sophisticated" tactics are too slow!!
+  have s0: fin.succ (0: fin 3) = 1, {norm_fin,},
+  have s1: fin.succ (1: fin 3) = 2, {norm_fin,},
+  have s2: fin.succ (2: fin 3) = 3, {norm_fin,},
+  have a0 : ((-1:R)^(↑(0: fin 4))) = (1:R), {rw fin.coe_zero,rw pow_zero,},
+  have a1 : (-1:R)^(↑(1: fin 4)) = -1, {rw fin.coe_one,rw pow_one,},
+  have a2 : (-1:R)^(↑(2: fin 4)) = 1, {
+    rw neg_one_pow_eq_pow_mod_two, 
+    rw fin.coe_two, 
+    rw nat.mod_self, 
+    rw pow_zero,
+  },
+  have a3 : (-1:R)^(↑(3: fin 4)) = -1, {
+    rw neg_one_pow_eq_pow_mod_two, 
+    have : ↑(3: fin 4) = (3:ℕ), {refl,}, 
+    rw this, 
+    have : 3 % 2  = 1, by refl, 
+    rw this, 
+    rw pow_one,
+  },
```

Comment:
These CI errors are telling you to format things as
```suggestion
  have a3 : (-1:R)^(↑(3: fin 4)) = -1,
  { rw neg_one_pow_eq_pow_mod_two, 
    have : ↑(3: fin 4) = (3:ℕ), {refl,}, 
    rw this, 
    have : 3 % 2  = 1, by refl, 
    rw this, 
    rw pow_one },
```
which is the mathlib style

Reply:
Which part is in the error message is saying that?

```
Run ! (_target/deps/mathlib/scripts/lint-style.py $(find src -name '*.lean') || true) | (grep -vE 'ERR_COP|ERR_LIN' || true) | grep '^'
Error: ERR_MOD: Module docstring missing, or too late
Warning: WRN_BRC: Probable misformatting of curly braces
Warning: WRN_BRC: Probable misformatting of curly braces
Warning: WRN_BRC: Probable misformatting of curly braces
Warning: WRN_BRC: Probable misformatting of curly braces
Warning: WRN_BRC: Probable misformatting of curly braces
Warning: WRN_BRC: Probable misformatting of curly braces
Warning: WRN_BRC: Probable misformatting of curly braces
Warning: WRN_BRC: Probable misformatting of curly braces
Warning: WRN_BRC: Probable misformatting of curly braces
Warning: WRN_BRC: Probable misformatting of curly braces
Error: Process completed with exit code 1.
```

What is ERR_MOD??

Reply:
Also can I run the linter locally? Is there an extension or some cmd tool?

Reply:
Never mind found it:

```
python .\_target\deps\mathlib\scripts\lint-style.py .\src\matrix_cookbook\1_basic.lean
::error file=src\matrix_cookbook\1_basic.lean,line=27,code=ERR_LIN::ERR_LIN: Line has more than 100 characters
::error file=src\matrix_cookbook\1_basic.lean,line=31,code=ERR_LIN::ERR_LIN: Line has more than 100 characters   
::error file=src\matrix_cookbook\1_basic.lean,line=34,code=ERR_LIN::ERR_LIN: Line has more than 100 characters   
::error file=src\matrix_cookbook\1_basic.lean,line=35,code=ERR_LIN::ERR_LIN: Line has more than 100 characters   
::error file=src\matrix_cookbook\1_basic.lean,line=43,code=ERR_LIN::ERR_LIN: Line has more than 100 characters   
::error file=src\matrix_cookbook\1_basic.lean,line=47,code=ERR_LIN::ERR_LIN: Line has more than 100 characters   
::error file=src\matrix_cookbook\1_basic.lean,line=53,code=ERR_LIN::ERR_LIN: Line has more than 100 characters   
::error file=src\matrix_cookbook\1_basic.lean,line=58,code=ERR_LIN::ERR_LIN: Line has more than 100 characters   
::error file=src\matrix_cookbook\1_basic.lean,line=68,code=ERR_LIN::ERR_LIN: Line has more than 100 characters   
::error file=src\matrix_cookbook\1_basic.lean,line=93,code=ERR_LIN::ERR_LIN: Line has more than 100 characters   
::error file=src\matrix_cookbook\1_basic.lean,line=99,code=ERR_LIN::ERR_LIN: Line has more than 100 characters   
::error file=src\matrix_cookbook\1_basic.lean,line=1,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\1_basic.lean,line=2,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\1_basic.lean,line=3,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\1_basic.lean,line=4,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\1_basic.lean,line=5,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\1_basic.lean,line=6,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\1_basic.lean,line=7,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\1_basic.lean,line=0,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\1_basic.lean,line=0,code=ERR_COP::ERR_COP: Malformed or missing copyright header 
::error file=src\matrix_cookbook\1_basic.lean,line=10,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\1_basic.lean,line=0,code=ERR_COP::ERR_COP: Malformed or missing copyright header 
::error file=src\matrix_cookbook\1_basic.lean,line=12,code=ERR_COP::ERR_COP: Malformed or missing copyright header

python .\_target\deps\mathlib\scripts\lint-style.py .\src\matrix_cookbook\mat_eq27_lib.lean
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=121,code=ERR_LIN::ERR_LIN: Line has more than 100 characters
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=123,code=ERR_LIN::ERR_LIN: Line has more than 100 characters  
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=124,code=ERR_LIN::ERR_LIN: Line has more than 100 characters  
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=125,code=ERR_LIN::ERR_LIN: Line has more than 100 characters  
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=126,code=ERR_LIN::ERR_LIN: Line has more than 100 characters  
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=127,code=ERR_LIN::ERR_LIN: Line has more than 100 characters  
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=128,code=ERR_LIN::ERR_LIN: Line has more than 100 characters  
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=129,code=ERR_LIN::ERR_LIN: Line has more than 100 characters  
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=162,code=ERR_LIN::ERR_LIN: Line has more than 100 characters  
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=163,code=ERR_LIN::ERR_LIN: Line has more than 100 characters  
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=164,code=ERR_LIN::ERR_LIN: Line has more than 100 characters  
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=1,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=2,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=3,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=4,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=5,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=6,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=7,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=8,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=0,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=10,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=10,code=ERR_MOD::ERR_MOD: Module docstring missing, or too late
::warning file=src\matrix_cookbook\mat_eq27_lib.lean,line=80,code=WRN_BRC::WRN_BRC: Probable misformatting of curly braces
::warning file=src\matrix_cookbook\mat_eq27_lib.lean,line=85,code=WRN_BRC::WRN_BRC: Probable misformatting of curly braces
::warning file=src\matrix_cookbook\mat_eq27_lib.lean,line=86,code=WRN_BRC::WRN_BRC: Probable misformatting of curly braces
::warning file=src\matrix_cookbook\mat_eq27_lib.lean,line=93,code=WRN_BRC::WRN_BRC: Probable misformatting of curly braces
::warning file=src\matrix_cookbook\mat_eq27_lib.lean,line=96,code=WRN_BRC::WRN_BRC: Probable misformatting of curly braces
::warning file=src\matrix_cookbook\mat_eq27_lib.lean,line=100,code=WRN_BRC::WRN_BRC: Probable misformatting of curly braces
::warning file=src\matrix_cookbook\mat_eq27_lib.lean,line=200,code=WRN_BRC::WRN_BRC: Probable misformatting of curly braces
::warning file=src\matrix_cookbook\mat_eq27_lib.lean,line=204,code=WRN_BRC::WRN_BRC: Probable misformatting of curly braces
::warning file=src\matrix_cookbook\mat_eq27_lib.lean,line=205,code=WRN_BRC::WRN_BRC: Probable misformatting of curly braces
::warning file=src\matrix_cookbook\mat_eq27_lib.lean,line=212,code=WRN_BRC::WRN_BRC: Probable misformatting of curly braces
```

Reply:
Note that in CI I'm ignoring `ERR_COP` and `ERR_LIN`